File: problems/142-Linked-List-Cycle-II/solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *detectCycle(struct ListNode *head) {
    struct ListNode *current = head;
    int len = 0;
    
    while(current) {
        if(current->val >= 900000) {
            break;
        }
        len++;
        current->val += 1000000;
        current = current->next;
    }
    
    current = head;
    for(int i = 0; i < len; i++) {
        current->val -= 1000000;
        current = current->next;
    }
    
    return current;
}

File: problems/142-Linked-List-Cycle-II/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func detectCycle(head *ListNode) *ListNode {
	slowPtr, fastPtr := head, head

	for fastPtr != nil && fastPtr.Next != nil {
		slowPtr = slowPtr.Next
		fastPtr = fastPtr.Next.Next

		if slowPtr == fastPtr {
			headPtr, collisionPtr := head, slowPtr
			for headPtr != collisionPtr {
				headPtr = headPtr.Next
				collisionPtr = collisionPtr.Next
			}

			return collisionPtr
		}
	}

	return nil
}

